experimental/dds/tree2/docs/Shared Tree Philosophy.md

```diff
@@ -0,0 +1,199 @@
+
+When users edit documents, those high level semantic operations need to be encoded in a way that supports merges.
+Additionally this encoding needs to be deterministically applied in all clients, even if they are using different version of Shared Tree.
+This needs to be done such that when desired, applications can ensure invariants hold.
+This means the editing and merge logic must be kept identical for compatibility, and changes to it instead have to be strictly new cases being supported:
+for example new edit operations can be added or new configuration flags added to existing ones can be added, but how a particular merge gets resolved can not be changed without introducing a compatibility issue.
+This can get complicated, so to keep it manageable, this problem was subdivided in a few different ways.
+
+The data-model for Shared Tree was selected to assist with subdividing this problem.
+The selected data-model consists of many small structures where collaborative edits interact—fields—, and nodes which connect them into a hierarchy and assigns types to each of the fields.
+
+The nodes have schema which define the schema for their fields.
+There are a few [kinds](<https://en.wikipedia.org/wiki/Kind_(type_theory)>) of nodes configure their fields differently, for example like structs (with a fixed set of mixed field types) or maps (with an extensible set of matching field types).
+
+There are also a few kinds of fields which provide different collaborative data-structures (like `sequence`, or `optional`), each of which hold nodes.
+
+All of the editing is done in the fields via the `field kind`.
+This allows the editing logic, both API and merge policy, to be packaged together into minimal versionable units ( `field kinds`).
+Since the schema explicitly selects what field kinds to use, updating of them is opt in, and can be coordinated with deployment schedules to achieve the required cross version collaboration requirements.
+This also leverages all the same collaboration as the rest od schema evolution, making adopting alternative field kinds no different from updating other aspects of the application's schema.
+
+This results in a tree where the overall shape is controlled by the application via its tree schema (for the nodes), but its made of alternating layers of fields provide all the complex logic, neatly separating the concerns.
+This can also be thought of as a tree of entities alternating between fields and nodes.
+As these two kinds of entities alternate and thus never touch, this resembles a [Bipartite Graph](https://en.wikipedia.org/wiki/Bipartite_graph), and can be called a Bipartite Tree.
+
+# Constraints and Sub-Trees
+
+When users edit documents, those high level semantic operations need to be encoded in a way that supports merges.
+Applications need to be able to easily ensure the invariants they require will be maintained across merged, and (possible with some effort) high fidelity merge resolution should be achievable.
```

Comment:
```suggestion
Applications need to be able to easily ensure the invariants they require will be maintained across merges.
```

Reply:
Not clear if the second part of the sentence adds anything (and maybe contradicts the first part?)

Reply:
It add something very important: We need to define success such that all changes are conflicted and rejected is considered bad and we preserve as much of edits as possible. Yes, this is in direct opposition to maintaining invariants, and balancing this in a way that the app is happy with is our job.

Reply:
I have clarified this in the rewriting of this section